Source/WTF/wtf/threads/Signals.cpp

```diff
@@ -222,6 +231,93 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+        // If the faulting address is out of the range of any valid memory, we would
+        // not have any reason to handle it. Just let the default handler take care of it.
+        static constexpr unsigned validAddressBits = OS_CONSTANT(EFFECTIVE_ADDRESS_WIDTH);
+        static constexpr uintptr_t invalidAddressMask = ~((1ull << validAddressBits) - 1);
+        if (bitwise_cast<uintptr_t>(info.faultingAddress) & invalidAddressMask)
+            return KERN_FAILURE;
+#endif
+    }
+
+    handlers.forEachHandler(signal, [&] (const SignalHandler& handler) {
+        SignalAction handlerResult = handler(signal, info, registers);
+        didHandle |= handlerResult == SignalAction::Handled;
+    });
+    return KERN_SUCCESS;
+}
+
+#if defined(EXCEPTION_STATE_IDENTITY_PROTECTED)
+
+kern_return_t catch_mach_exception_raise_state_identity_protected(
+    mach_port_t exceptionPort,
+    __unused uint64_t threadID,
+    __unused mach_port_t taskIDToken,
+    exception_type_t exceptionType,
+    mach_exception_data_t exceptionData,
+    mach_msg_type_number_t dataCount,
+    int* stateFlavor,
+    const thread_state_t inState,
+    mach_msg_type_number_t inStateCount,
+    thread_state_t outState,
+    mach_msg_type_number_t* outStateCount)
+{
+    SignalHandlers& handlers = g_wtfConfig.signalHandlers;
+    RELEASE_ASSERT(exceptionPort == handlers.exceptionPort);
+    // If we wanted to distinguish between SIGBUS and SIGSEGV for EXC_BAD_ACCESS on Darwin we could do:
+    // if (exceptionData[0] == KERN_INVALID_ADDRESS)
+    //    signal = SIGSEGV;
+    // else
+    //    signal = SIGBUS;
+    Signal signal = fromMachException(exceptionType);
+    RELEASE_ASSERT(signal != Signal::Unknown);
+
+#if CPU(ARM64E) && OS(DARWIN)
+    ptrauth_generic_signature_t inStateHash = hashThreadState(inState);
+#endif
+
+    memcpy(outState, inState, inStateCount * sizeof(inState[0]));
+
+#if CPU(X86_64)
+    RELEASE_ASSERT(*stateFlavor == x86_THREAD_STATE);
+    PlatformRegisters& registers = reinterpret_cast<x86_thread_state_t*>(outState)->uts.ts64;
+#elif CPU(X86)
+    RELEASE_ASSERT(*stateFlavor == x86_THREAD_STATE);
+    PlatformRegisters& registers = reinterpret_cast<x86_thread_state_t*>(outState)->uts.ts32;
+#elif CPU(ARM64)
+    RELEASE_ASSERT(*stateFlavor == ARM_THREAD_STATE);
+    PlatformRegisters& registers = reinterpret_cast<arm_unified_thread_state*>(outState)->ts_64;
+#elif CPU(ARM)
+    RELEASE_ASSERT(*stateFlavor == ARM_THREAD_STATE);
+    PlatformRegisters& registers = reinterpret_cast<arm_unified_thread_state*>(outState)->ts_32;
+#endif
+
+    bool didHandle = false;
+    kern_return_t kr = runSignalHandlers(signal, registers, didHandle, dataCount, exceptionData);
+    if (kr != KERN_SUCCESS)
+        return kr;
+
+    if (didHandle) {
+#if CPU(ARM64E) && OS(DARWIN)
+        RELEASE_ASSERT(inStateHash == hashThreadState(outState));
+#endif
+        *outStateCount = inStateCount;
+        return KERN_SUCCESS;
+    }
+
+    return KERN_FAILURE;
```

Comment:
This looks identical to `catch_mach_exception_raise_state`. Can we just push this all into one shared function?